### PR TITLE
Create hmm_poisson_changepoint_jax.ipynb

### PR DESCRIPTION
**Pull request by karalleyna**
### Figures
image
image
image
image

## Checklist:

- [x] Performed a self-review of the code
- [x] Tested on Google Colab.

### CODE REVIEW

**Comment by murphyk** [comment]
This is confusing:
```
 trans_dist=distrax.Categorical(probs=transition_probs),
    init_dist=distrax.Categorical(logits=initial_state_logits))
```
I suggest making init_dist be probabilities instead of logits (even if it does not match tfp demo).



I would rename ` get_hmm(trainable_log_rates, transition_probs, initial_state_logits):` to be `make_hmm`.

if you do `!git clone https://github.com/probml/pyprobml /pyprobml &> /dev/null`It clones the entire repo so you don't need `!wget -q https://raw.githubusercontent.com/probml/pyprobml/master/scripts/hmm_lib.py`

This is rather complex:
```
def build_latent_state(num_states, max_num_states, daily_change_prob):
  # Give probability exp(-100) ~= 0 to states outside of the current model.
  def logit(s):
    return jnp.where(s < num_states + 1, 0., -100.)
 
  states = jnp.arange(1, max_num_states+1)
  initial_state_logits = vmap(logit)(states)
 
  # Build a transition matrix that transitions only within the current
  # `num_states` states.
  def transition_prob(i, s):
      return jnp.where((s <= num_states) & (i <= num_states) & (num_states>1), 
                      jnp.where(s == i, 1-daily_change_prob, daily_change_prob/(num_states-1)), 
                      jnp.where(s == i, 1, 0))
      
  transition_probs = vmap(transition_prob, in_axes=(None, 0))(states, states)
 ```

I think you can use the tfp solution below which is easier to read. Although it uses numpy destructively (fill_diagonal is an assignment operator), it is inside the function call, so it is still "functionally pure" - it is just an easy way to create the initial static jax.numpy arrays. (ie you can use np.foo to create initial values that are assigned to jax, as long as it is deterministic).
```

def build_latent_state(num_states, max_num_states, daily_change_prob=0.05):

  # Give probability exp(-100) ~= 0 to states outside of the current model.
  initial_state_logits = -100. * np.ones([max_num_states], dtype=np.float32)
  initial_state_logits[:num_states] = 0.

  # Build a transition matrix that transitions only within the current
  # `num_states` states.
  transition_probs = np.eye(max_num_states, dtype=np.float32)
  if num_states > 1:
    transition_probs[:num_states, :num_states] = (
        daily_change_prob / (num_states-1))
    np.fill_diagonal(transition_probs[:num_states, :num_states],
                     1-daily_change_prob)
  return initial_state_logits, transition_probs
```

**Comment by karalleyna** [comment]
> I think you can use the tfp solution below which is easier to read. Although it uses numpy destructively (fill_diagonal is an assignment operator), it is inside the function call, so it is still "functionally pure" - it is just an easy way to create the initial static jax.numpy arrays. (ie you can use np.foo to create initial values that are assigned to jax, as long as it is deterministic).
> 
> ```
> 
> def build_latent_state(num_states, max_num_states, daily_change_prob=0.05):
> 
>   # Give probability exp(-100) ~= 0 to states outside of the current model.
>   initial_state_logits = -100. * np.ones([max_num_states], dtype=np.float32)
>   initial_state_logits[:num_states] = 0.
> 
>   # Build a transition matrix that transitions only within the current
>   # `num_states` states.
>   transition_probs = np.eye(max_num_states, dtype=np.float32)
>   if num_states > 1:
>     transition_probs[:num_states, :num_states] = (
>         daily_change_prob / (num_states-1))
>     np.fill_diagonal(transition_probs[:num_states, :num_states],
>                      1-daily_change_prob)
>   return initial_state_logits, transition_probs
> ```

If I use the original code,  I cannot use vmap to create models with different number of hidden states like

```
# For each candidate model, build the initial state prior and transition matrix.
batch_initial_state_logits, batch_transition_probs = vmap(build_latent_state, in_axes=(0, None, None))(states, max_num_states, daily_change_prob)
```
rather than 

```
# For each candidate model, build the initial state prior and transition matrix.
batch_initial_state_logits = []
batch_transition_probs = []
for num_states in range(1, max_num_states+1):
  initial_state_logits, transition_probs = build_latent_state(
      num_states=num_states,
      max_num_states=max_num_states)
  batch_initial_state_logits.append(initial_state_logits)
  batch_transition_probs.append(transition_probs)
```

So, should build_latent_space be kept as it is?

**Comment by murphyk** [comment]
Oh, I see. You can keep your build_latent_space as is then.

**Comment by karalleyna** [comment]
Done 🙂
